pkg/controller/atlasprivateendpoint/privateendpoint.go

```diff
@@ -0,0 +1,220 @@
+		}
+	case !existInAtlas && wasDeleted:
+		return r.unmanage(ctx, akoPrivateEndpoint)
+	case existInAtlas && wasDeleted:
+		if customresource.IsResourcePolicyKeepOrDefault(akoPrivateEndpoint, r.ObjectDeletionProtection) {
+			return r.unmanage(ctx, akoPrivateEndpoint)
+		}
+
+		atlasPEService, err = r.deletePrivateEndpoint(ctx.Context, projectID, atlasPEService)
+		if err != nil {
+			return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointFailedToDelete, err)
+		}
+	}
+
+	switch atlasPEService.Status() {
+	case privateendpoint.StatusInitiating:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is being initialized")
+	case privateendpoint.StatusPending, privateendpoint.StatusPendingAcceptance, privateendpoint.StatusWaitingForUser, privateendpoint.StatusVerified:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceInitializing, "Private Endpoint is waiting for human action")
+	case privateendpoint.StatusFailed, privateendpoint.StatusRejected:
+		return r.terminate(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceFailedToConfigure, errors.New(atlasPEService.ErrorMessage()))
+	case privateendpoint.StatusDeleting:
+		return r.inProgress(ctx, akoPrivateEndpoint, atlasPEService, api.PrivateEndpointServiceReady, workflow.PrivateEndpointServiceDeleting, "Private Endpoint is being deleted")
+	}
+
+	ctx.SetConditionTrue(api.PrivateEndpointServiceReady)
+	r.EventRecorder.Event(akoPrivateEndpoint, "Normal", string(workflow.PrivateEndpointServiceCreated), "Private Endpoint Service is available")
+
+	return r.handlePrivateEndpointInterface(ctx, projectID, akoPrivateEndpoint, akoPEService, atlasPEService)
+}
+
+func (r *AtlasPrivateEndpointReconciler) handlePrivateEndpointInterface(
+	ctx *workflow.Context,
+	projectID string,
+	akoPrivateEndpoint *akov2.AtlasPrivateEndpoint,
+	akoPEService privateendpoint.EndpointService,
+	atlasPEService privateendpoint.EndpointService,
+) (ctrl.Result, error) {
+	if len(akoPEService.EndpointInterfaces()) == 0 && len(atlasPEService.EndpointInterfaces()) == 0 {
+		return r.waitForConfiguration(ctx, akoPrivateEndpoint, atlasPEService)
+	}
+
+	inProgress := false
+	for _, akoPEInterface := range akoPEService.EndpointInterfaces() {
+		atlasPEInterfaces := atlasPEService.EndpointInterfaces()
+		atlasPEInterface := atlasPEInterfaces.Get(akoPEInterface.InterfaceID())
+		existInAtlas := atlasPEInterface != nil
+		inProgress = isInterfaceInProgress(akoPEInterface.Status())
+		var err error
+
+		if !existInAtlas {
+			gcpProjectID := getGCPProjectID(akoPrivateEndpoint, akoPEInterface.InterfaceID())
```

Comment:
Here is because of Atlas (again). We need this to configure a GCP "interface" but Atlas doesn't give this field back to us.
We send it when creating, but when getting the resource we don't have this field, so we always send it to the translation layer even when it is empty.

Reply:
Ok, that is weird alright, but still the original question is, why is this code **not** guarded to run only when the provider is Google?

Reply:
Because there are no separate flows per provider, there's a creation flow and all the data should be there depending on the provider.

Reply:
Ok I see the guard inside the function:
```go
func getGCPProjectID(...) string {
	if akoPrivateEndpoint.Spec.Provider == privateendpoint.ProviderGCP && len(akoPrivateEndpoint.Spec.GCPConfiguration) > 0 {
...
	}

	return ""
}
```